runtime/bcverify/rtverify.c

```diff
@@ -1165,15 +1165,50 @@ verifyBytecodes (J9BytecodeVerificationData * verifyData)
 				break;
 
 			case JBreturn1:
+			case JBreturnB:
+			case JBreturnC:
+			case JBreturnS:
+			case JBreturnZ:
 			case JBsyncReturn1:
 				if (returnChar) {
+					U_32 returnType = 0;
 					/* single character return description */
 					if ((returnChar < 'A') || (returnChar > 'Z')) {
 						inconsistentStack = TRUE;
 						goto _illegalPrimitiveReturn;
 					}
 					type = (UDATA) oneArgTypeCharConversion[returnChar - 'A'];
 					POP_TOS(type);
+
+					/* check methods that return char, byte, short, or bool use the right opcode */
+					if (BCV_BASE_TYPE_INT == type){
+						switch(bc){
+						case JBreturnB:
+							returnType = BCV_BASE_TYPE_BYTE_BIT;
+							break;
+
+						case JBreturnC:
+							returnType = BCV_BASE_TYPE_CHAR_BIT;
+							break;
+
+						case JBreturnS:
+							returnType = BCV_BASE_TYPE_SHORT_BIT;
+							break;
+
+						case JBreturnZ:
+							returnType = BCV_BASE_TYPE_BOOL_BIT;
+							break;
+
+						case JBreturn1:
+						case JBsyncReturn1:
+							returnType = BCV_BASE_TYPE_INT_BIT;
+							break;
+						default:
+							//TODO mark unreachable
+							break;
+						}
+						inconsistentStack |= ( returnType == baseTypeCharConversion[returnChar - 'A']);
```

Comment:
Does this match the RIs behaviour?  Some of the primitives can be automatically converted to others as long as the conversion doesn't lose info.  ie: byte -> int.  Will this fail verification in those cases?

Reply:
Does this actually matter, since these return bytecodes are only generated internally?

Reply:
I think you're right here - the check that the return type is compatible is already handled above this code so all this is doing is validating we put the right bytecode in for the signature.